o1vm/src/mips/constraints.rs

```diff
@@ -415,65 +416,75 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
             self.variable(Self::Position::ScratchState(MIPS_HAS_N_BYTES_OFF + i))
         });
 
-        // TODO: any constraints we should we add for pos?
+        // The actual number of bytes read in this instruction, will be 0 <= x <= len <= 4
+        let actual_read_bytes = self.variable(pos);
 
         // EXTRA 13 CONSTRAINTS
 
-        // Booleanity constraints
+        // 5 Booleanity constraints
         {
             for var in has_n_bytes.iter() {
                 self.assert_boolean(var.clone());
             }
             self.assert_boolean(end_of_preimage.clone());
         }
 
+        // + 4 constraints
         {
             // Expressions that are nonzero when the exact corresponding number
-            // of bytes are read (case 0 bytes used when bytelength is read)
-            // TODO: use equal?
-            let read_1 = (num_read_bytes.clone())
-                * (num_read_bytes.clone() - Expr::from(2))
-                * (num_read_bytes.clone() - Expr::from(3))
-                * (num_read_bytes.clone() - Expr::from(4));
-            let read_2 = (num_read_bytes.clone())
-                * (num_read_bytes.clone() - Expr::from(1))
-                * (num_read_bytes.clone() - Expr::from(3))
-                * (num_read_bytes.clone() - Expr::from(4));
-            let read_3 = (num_read_bytes.clone())
-                * (num_read_bytes.clone() - Expr::from(1))
-                * (num_read_bytes.clone() - Expr::from(2))
-                * (num_read_bytes.clone() - Expr::from(4));
-            let read_4 = (num_read_bytes.clone())
-                * (num_read_bytes.clone() - Expr::from(1))
-                * (num_read_bytes.clone() - Expr::from(2))
-                * (num_read_bytes.clone() - Expr::from(3));
-
-            // Note these constraints also hold when 0 preimage bytes are read
+            // of preimage bytes are read (case 0 bytes used when bytelength is read)
+            // TODO: embed any more complex logic to know how many bytes are read
+            //       depending on the address and length as in the witness?
+            // FIXME: use the lines below when the issue with `equal` is solved
```

Comment:
I do not understand what you mean with "when the issue with 'equal' is solved". The commented lines below will never be satisfied all at the same time, as only one will be true.

Reply:
I meant that the lines using `equal` are equivalent but cleaner than the lines using the explicit values with multiplications. Nonetheless, they were failing: I know only one will be true and the rest won't, same as with the multiplication values. What I meant is that the constraints using `preimage_n` below fail when `preimage_n` is defined with `equal` instead of the multiplications approach. And I believe we should express them with `equal` in the future and fix that issuel.